Variadic ODE design doc (design-doc #19)

bbbales2 opened the pull request:
Rendered version [here](https://github.com/stan-dev/design-docs/blob/feature/variadic_odes/designs/0019-variadic_odes.md)

There's a bit of code here that has implemented most of this design: https://github.com/stan-dev/math/pull/1641

bbbales2 commented:
@wds15, @yizhang-yiz, @charlesm93, @bob-carpenter -- whoever wants to comment or review, have at it. I'd like to get something like this through in the release 3 months from now.

wds15 commented:
Really cool to see that much action from you on this! I will take a deep dive on this soon.

bbbales2 commented:
@wds15 I'm bored and haven't coded stuff in a long time. Gimme something to do.

wds15 commented:
Sorry for the long delay...

Do we have benchmarks new vs old stuff?

wds15 commented on designs/0019-variadic_odes.md:
```diff
@@ -0,0 +1,173 @@
+that could otherwise be passed to through as a right hand side argument here).
+
+I'm not opposed to this either, and this also isn't incompatible with the
+current proposal, though it would probably require some extra work.
```
Na, we should not wait here. This is great stuff and will make things easier for users. It may actually make things faster as fewer copies will be needed due to avoiding the unpacking step in the RHS. Moreover, we also avoid that *unpacked data* will be cast to a `var` within the ODE RHS, which is probably what happens at the moment for more complicated models.

bbbales2 replied:
Weeeeee

wds15 commented on designs/0019-variadic_odes.md:
````diff
@@ -0,0 +1,173 @@
+2. ```y0``` - Initial state of the ode solve (`y0 = y(t0)`)
+3. ```t0``` - Initial time of the ode solve
+4. ```times``` - Sequences of times to which the ode will be solved (each
+  element must be greater than t0).
````
In the current implementation, the given times must be a strictly increasing vector. It would be far more convenient if the times vector must only be an increasing vector, but adjacent times are allowed to be repeated, so this should be legal for times:

1, 2, 3.2, 3.2, 4

This has caused quite some trouble for users - including myself. Can we allow that?

bbbales2 replied:
Good suggestion. Yes we can change this. Should we allow t = 0.0 to be in the times as well? I don't see why not.

wds15 replied:
Sure, lets allow that as well.

bbbales2 replied:
I ended up not allowing for t = t0 because I got confused thinking about how it would interact with t0/y0.

bbbales2 commented on designs/0019-variadic_odes.md:
````diff
@@ -0,0 +1,173 @@
+
+The proposed `ode_bdf` interface is:
+```
+vector[] ode_bdf(F f,
````
ODE math solver-side is usually written in terms of vectors. The state of the ode is a vector. And I was writing some ODE solvers in Stan and it was annoying to put to_array/to_vector everywhere to keep things straightened out.

bbbales2 commented on designs/0019-variadic_odes.md:
```diff
@@ -0,0 +1,173 @@
+  side (`f`). Can be any non-function type.
+
+There is no way to optionally only provide some of the `rel_tol`, `abs_tol`, or
+`max_num_steps` arguments. We can keep the current defaults (`1e-6`, `1e-6`,
```
Looks like they are inconsistent:

https://github.com/stan-dev/math/blob/develop/stan/math/rev/functor/integrate_ode_bdf.hpp

https://github.com/stan-dev/math/blob/develop/stan/math/prim/functor/integrate_ode_rk45.hpp

https://mc-stan.org/docs/2_23/stan-users-guide/control-parameters-for-ode-solving.html

bbbales2 commented on designs/0019-variadic_odes.md:
```diff
@@ -0,0 +1,173 @@
+(with no rules on naming).
+2. The state is written in terms of `vector`s instead of real arrays. ODE states
+are mathematically written and expressed as vectors, so given the interface is
+changing we might as well change this to be consistent.
```
Good point ODEs are usually defined element by element. But they are still written as elements of vectors.

bbbales2 commented on designs/0019-variadic_odes.md:
```diff
@@ -0,0 +1,173 @@
+we'd make it a vector argument (so it wasn't like one argument of `ode_bdf`
+needs to be a vector and one doesn't). However, then it seems like the output
+of `ode_bdf` would be a matrix. I think the output makes more sense as an array
+of vectors.
```
Which aspect?

wds15 replied:
My thought was that in Stan the array of vector thing is the natural container for multi-variate stuff... and the algebra solver uses the same conventions.

bbbales2 commented on designs/0019-variadic_odes.md:
```diff
@@ -0,0 +1,173 @@
+vector[] ode_bdf_tol(F f,
+                     vector y0,
+                     real t0, vector times,
+                     real rel_tol, real abs_tol,
```
> From what I have seen it can be beneficial to have more accuracy for the state, which you can relax the accuracy for the gradients.

It would be nice to have optional arguments. I kinda lean yes to exposing things. Without doing that we won't be able to experiment either. This will probably also change with the adjoint stuff. I think something for later.

wds15 replied:
So, can you expose that functionality from CVODES so that we can test it a bit?

bbbales2 replied:
Yeah sure, we can try it

bbbales2 commented on designs/0019-variadic_odes.md:
```diff
@@ -0,0 +1,173 @@
+[guide-level-explanation]: #guide-level-explanation
+
+For the bdf solver, I am proposing introducing two new functions: `ode_bdf` and
+`ode_tol_bdf` (there will be two functions similarly introduced for the rk45 and
```
Doesn't bother me. We're just getting into crazy long name territory. MATLAB has super short names like ode45 etc.

I think integrate/solve are a bit extraneous. @bob-carpenter I keep calling you in on naming things. What do you think here?

`ode_bdf`, `solve_ode_bdf`, `integrate_ode_bdf_2`? They're all fine to me.

wds15 replied:
Having @bob-carpenter‘s view on this would be nice, indeed

bbbales2 commented:
> Do we have benchmarks new vs old stuff?

I do not. I'm just assuming this interface can be fast enough, and then we can check it before it merges that it indeed is.

bbbales2 commented:
Eek, was there something that you wanted me to write up here (said something about it in the meeting)?

I just started writing code. I guess things I've done are:

1. Converted to doing state as an Eigen::VectorXd
2. Written a bunch of docs and added more tests
3. Fixed the t0 issue again (hopefully for good this time)
4. And now I've started adding in the adjoint stuff in, which isn't really part of this design doc, but I'm doing it in the ODE thread. We might break that up into separate things -- I kinda just saw the opportunity and started.

wds15 commented:
I was expecting to hear back so that we start exploring those additional tolerance control things for example. We still need to decide on what options to control in addition of any.

Btw, as we seem to switch to general template arguments you can do that here now as well which will probably make the code more memory efficient since some copies can be avoided, I think.

bbbales2 commented:
> I was expecting to hear back so that we start exploring those additional tolerance control things for example

Oh yeah, working through the interface now. We'll go back and forth on that once I get a baseline up.

wds15 commented:
Just as a note: The just released sundial 5.3 includes in the release notes: "Our latest release, v5.3.0, adds support to CVODE for integrating IVPs with contstraints and projecting the solution onto the constraint manifold"

Sounds cool. 

https://computing.llnl.gov/projects/sundials/sundials-software

totally not for this design per se, but could be a consideration should we want to move into this direction in the future.

bbbales2 commented:
Oh interesting. That could be an easier way to handle positivity constraints.

But then we couldn't easily expose that functionality to the other solvers.

bbbales2 commented:
I think the basic code and such is in place to evaluate this and start wrapping everything up.

There are three outstanding questions I know of (it's possible more get added in the course of review), but let me tag some people and see if we can get these out of the way.

The writeup is [here](https://github.com/stan-dev/design-docs/blob/feature/variadic_odes/designs/0019-variadic_odes.md).

1. We should decide on default tolerances. The docs are inconsistent now anyway. Currently CVODES solvers run at `rtol = 1e-10`, `atol = 1e-10` and the RK45 solver runs at `rtol = 1e-6`, `atol = 1e-6`.

2. We need to decide on the error control for CVODES (whether or not to include sensitivity variables in error control, see [here](https://github.com/stan-dev/math/pull/1641#issuecomment-635562506))

3. Naming. I originally did short names like `ode_bdf` for convenience. @wds15 likes `integrate_ode_bdf_2` ([details](https://github.com/stan-dev/design-docs/pull/20#discussion_r417816535)). Either one will do. What seems better from a user perspective?

@wds15 @charlesm93 @yizhang-yiz @bob-carpenter

wds15 commented:
1. Is it an option to drop for the new interfaces the functions which set default tolerances? I do think that people should by all means set the tolerances.

2. I hope my benchmarks will be done next week and give some help to decide.

3. I am fine with short names. So ode_bdf is great for me...just like matlab

bbbales2 commented:
@wds15 let's keep default tolerances so-as to not rock the boat. I'll copy them over and fix the doc to reflect the difference if nobody has any other suggestions.

I also like the different names since the interface is changing slightly (vector initial conditions + tolerances in different places). The new solvers aren't just a drop in substitute for the old solvers (though that is handled via the backwards compatibility `integrate_ode_bdf` etc. interfaces here).

bbbales2 commented:
1. I've done BDF/Adams defaults as `rtol = 1e-10`, `atol = 1e-10` and RK45 solver defaults at `rtol = 1e-6`, `atol = 1e-6`.

2. I've configured CVODES (BDF/Adams) to adjust timestep based on tolerance requirements of the sensitivity equations.

3. Short names are good.

wds15 commented on designs/0019-variadic_odes.md:
```diff
@@ -0,0 +1,171 @@
+code from the
+[reduce_sum](https://github.com/stan-dev/design-docs/blob/master/designs/0017-reduce_sum.md)
+design doc and the existing ODE infrastructure. There is not any significant new
+technology needed.
```
Maybe mention that we now also need a `zero_adjoints` utility function which wasn't required by `reduce_sum` before.

wds15 commented on designs/0019-variadic_odes.md:
```diff
@@ -0,0 +1,171 @@
+
+One drawback is that the existing ODE documentation and case studies will need
+to be updated to point at the new interfaces. The old interfaces can
+remain for backwards compatibility though.
```
The other drawback we need to mention here is that the old `integrate*` functions will slightly slow down since we have to adapt things around which is noticeable with a few percent performance drop (most notably for RK45 where the adapter gets in the way a bit).

wds15 commented on designs/0019-variadic_odes.md:
```diff
@@ -0,0 +1,171 @@
+depend on.
+
+Argument packing is inconvenient, makes code hard to read, and introduces more
+chances for making errors.
```
Maybe add as advantage that the requirement for packing leads easily to over-propagation of packed data to properly named variables which are cast to `vars` which is avoided with the new variadic interface.

bbbales2 replied:
Can't conceptually we put all data in the x_r arguments so there's no promotion?

wds15 commented on designs/0019-variadic_odes.md:
```diff
@@ -0,0 +1,171 @@
+preferable). Also ODE states are usually represented as vectors mathematically.
+
+3. The times array only needs to be sorted, not strictly increasing. This means
+that there can multiple output times in a row with the same value.
```
We need to add somewhere here that we change the CVODES way of checking for errors to now include the sensitivities. This was done in response to experiments with Adams which got off the rails in our current implementation on rather simple examples (and this is avoided with the new checking scheme).

wds15 commented on designs/0019-variadic_odes.md:
```diff
@@ -0,0 +1,171 @@
+account when computing local errors
+(https://github.com/stan-dev/math/pull/1641#issuecomment-635562506).
+
+3. We should decide on the naming. I'm not personally commited to `ode_bdf`
```
before merging this, the wording should be less personal flavoured. We are going with `ode_bdf` is what I would say.

wds15 commented on designs/0019-variadic_odes.md:
```diff
@@ -0,0 +1,171 @@
+# Prior art
+[prior-art]: #prior-art
+
+This builds directly on the existing ODE solver and `reduce_sum` interfaces.
```
Mention that the old tests stay in place.

bbbales2 replied:
I put this in the reference level explanation.

bbbales2 commented:
@wds15 did everything but the promotion. We have answers for all the unanswered questions, so I just cleaned those up.

rok-cesnovar commented:
Should we now approve/merge this? Everything proposed here now lives in Stan, as of 2.24

wds15 commented:
I would suggest to merge, yes.

rok-cesnovar commented:
@wds15 want to approve? You have done most of the reviewing.

wds15 commented:
I just approved once more. Not sure why this is still roadblocked.

rok-cesnovar commented:
You seem to not have approving rights (gray tick on approval), let me try.

rok-cesnovar commented:
Someone with proper permissions will have to do it.

mitzimorris commented:
maintainers only have read access.  people with write access are BB, Carp, Steve Bronder, and D Lee.  an odd set of folks.  Steve might have been given write access in order to manage the CoC issue which isn't really a design doc.
choices:  grant write access to all maintainers; get approval from BB, Steve, Daniel, or Bob.
